### PR TITLE
Calculating reward distribution

### LINKED ISSUE #107: Implement payout script

### LINKED ISSUE #108: Add **payout** CLI command

### LINKED ISSUE #125: Implement reward distribution function

### PR DESCRIPTION
**Pull request by MakMuftic**
**Implement logic for calculating reward distribution based on node stats (part of payout script). Some additional minor changes described below**

### PR Checklist

- [x] I have run linter localy
- [x] I have run unit and integration tests locally
- [x] Rebased to master branch / merged master
- [x] Updated CHANGELOG.md

### Changes

- Refactor stats endpoint to return payout addresses for nodes and loadbalancer fee percentage
- Create initial (empty) payout on the first loadbalancer start
- Add function for calculating reward distribution for nodes `CalculatePayoutDistributionByNode` + unit tests (@mpetrun5 @mpetrunic please check this thoroughly)
- Add function for aggregating information on payout stats + unit tests

### Issues

Closes #125

### CODE REVIEW

**Comment by MakMuftic** [comment]
@mpetrun5 @mpetrunic Failing CI is linked with [this](https://github.com/centrifuge/go-substrate-rpc-client/issues/107)

File: cmd/payout.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+		totalRewardAsFloat64,
+		float64(stats.Fee),
+	)
+
+	transactionDetails, err := payout.ExecuteAllPayoutTransactions(
+		distributionByNode,
+		stats.Stats,
+		secret,
+		loadbalancerUrl,
+	)
```
Comment (mpetrunic) [inline cmd/payout.go]: I think there should just be method `payout` which is executed by cmd with totalRewards and fee and it fetches stats, distribution and executes transaction.

File: cmd/payout.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	if err != nil {
+		log.Errorf("Unable to execute payout transactions, because of %v", err)
+		return
+	}
+
+	// todo - prettify displaying transactions status
+	log.Info(transactionDetails)
+}
+
+func fetchStatsFromEndpoint(endpoint string) (*controllers.StatsResponse, error) {
```
Comment (mpetrunic) [inline cmd/payout.go]: this shouldn't be in cmd

File: internal/payout/transaction.go [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+		}
+		if status.IsFinalized {
+			log.Debugf("Completed at block hash: %#x\n", status.AsFinalized)
+			return &TransactionDetails{
+				to:     to,
+				amount: amount,
+				status: Finalized,
+			}, nil
+		}
+		log.Debugf("Transaction status: %v#\n", status)
```
Comment (mpetrunic) [inline internal/payout/transaction.go]: I feel like tx status should be logged on info level

File: internal/payout/transaction.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+	}
+
+	// unlock segment
+	mux.Unlock()
+
+	defer sub.Unsubscribe()
+	for {
+		status := <-sub.Chan()
+		if status.IsDropped {
+			log.Debug("Dropped transaction")
```
Comment (mpetrunic) [inline internal/payout/transaction.go]: wanr, error?

**Review by mpetrunic — CHANGES_REQUESTED** [review]
@mpetrun5 Could probably review this beteer.

It would be easier to review if tx submit (rpc call) and reward calculation would be in separate PRs

**Comment by MakMuftic** [comment]
@mpetrun5 @mpetrunic I decided to separate changes in multiple PRs as suggested by @mpetrunic.

File: internal/controllers/stats.go [hunk 5]
```diff
@@ -2,19 +2,25 @@ package controllers
...
-	Stats map[string]models.NodeStatsDetails `json:"stats"`
+	Stats map[string]payout.NodePayoutDetails `json:"stats"`
+	Fee   float32                             `json:"fee"`
 }
 
+var getNow = time.Now
+
 func (c *ApiController) StatisticsHandlerAllStats(w http.ResponseWriter, r *http.Request) {
-	statisticsForPayout, err := stats.CalculateStatisticsFromLastPayout(c.repositories)
+	// should check for signature in body and only than record payout
```
Comment (mpetrunic) [inline internal/controllers/stats.go]:
```suggestion
	// should check for signature in body and only then record payout
```

File: internal/payout/distribution.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+	totalDistributedLivelinessRewards := float64(0)
+	totalDistributedRequestsRewards := float64(0)
+	payoutAmountDistributionByNodes := make(map[string]big.Int, len(payoutDetails))
+
+	for nodeId, nodeStatsDetails := range payoutDetails {
+		// liveliness rewards
+		nodeLivelinessRewardPercentage := nodeStatsDetails.TotalPings / totalNumberOfPings
+		roundedNodeLivelinessRewardPercentage :=
+			math.Floor(nodeLivelinessRewardPercentage*roundPrecision) / roundPrecision
+		livelinessReward := livelinessRewardPool * roundedNodeLivelinessRewardPercentage
```
Comment (mpetrunic) [inline internal/payout/distribution.go]: why do you need to round percentage?. Can't you just multiply by nodeLivelinessRewardPercentage and math.Floor it to make it round?

File: internal/payout/distribution.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+const (
+	livelinessRewardPercentage = 0.1
+	requestsRewardPercentage   = 0.9
+	roundPrecision             = 1_000_000_000_000_000
+)
+
+func CalculatePayoutDistributionByNode(
+	payoutDetails map[string]models.NodeStatsDetails,
+	totalReward float64,
+	loadBalancerFee float64,
```
Comment (mpetrun5) [inline internal/payout/distribution.go]: Maybe it makes more sense to call this loadBalancerFeePercentage to name it similar to livelinessReward and requestReward percentages and then call rewardPool * loadBalancerFeePercentage = loadbalancerReward

File: internal/payout/distribution.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+		totalNumberOfRequests += node.TotalRequests
+	}
+
+	totalDistributedLivelinessRewards := float64(0)
+	totalDistributedRequestsRewards := float64(0)
+	payoutAmountDistributionByNodes := make(map[string]big.Int, len(payoutDetails))
+
+	for nodeId, nodeStatsDetails := range payoutDetails {
+		// liveliness rewards
+		nodeLivelinessRewardPercentage := nodeStatsDetails.TotalPings / totalNumberOfPings
```
Comment (mpetrun5) [inline internal/payout/distribution.go]: Can totalNumberofPings be zero?

File: internal/payout/distribution.go (hunk 6 above)
Comment (MakMuftic) [reply internal/payout/distribution.go]: :+1:

File: internal/payout/distribution.go (hunk 7 above)
Comment (MakMuftic) [reply internal/payout/distribution.go]: :+1:

File: internal/payout/distribution.go (hunk 8 above)
Comment (MakMuftic) [reply internal/payout/distribution.go]: It shouldn't really happen but I will safe-guard it never the less

File: internal/payout/distribution.go (hunk 7 above)
Comment (MakMuftic) [reply internal/payout/distribution.go:17]: :+1:

**Review by mpetrun5 — APPROVED** [review]
LGTM

**Review by mpetrunic — APPROVED** [review]

**Pull request merged**